// next/dist/client/components/react-dev-overlay/ui/components/errors/dev-tools-indicator/dev-tools-indicator.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
0 && (module.exports = {
    DEV_TOOLS_INDICATOR_STYLES: null,
    DevToolsIndicator: null
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    DEV_TOOLS_INDICATOR_STYLES: function() {
        return DEV_TOOLS_INDICATOR_STYLES;
    },
    DevToolsIndicator: function() {
        return DevToolsIndicator;
    }
});
const _interop_require_default = require("@swc/helpers/_/_interop_require_default");
const _jsxruntime = require("react/jsx-runtime");
const _shared = require("../../../../shared");
const _react = require("react");
const _toast = require("../../toast");
const _nextlogo = require("./next-logo");
const _initialize = require("../../../../../../dev/dev-build-indicator/internal/initialize");
const _devrenderindicator = require("../../../../utils/dev-indicator/dev-render-indicator");
const _usedelayedrender = require("../../../hooks/use-delayed-render");
const _turbopackinfo = require("./dev-tools-info/turbopack-info");
const _routeinfo = require("./dev-tools-info/route-info");
const _gearicon = /*#__PURE__*/ _interop_require_default._(require("../../../icons/gear-icon"));
const _userpreferences = require("./dev-tools-info/user-preferences");
// TODO: add E2E tests to cover different scenarios
const INDICATOR_POSITION = process.env.__NEXT_DEV_INDICATOR_POSITION || 'bottom-left';
function DevToolsIndicator(param) {
    let { state, errorCount, isBuildError, setIsErrorOverlayOpen } = param;
    const [isDevToolsIndicatorVisible, setIsDevToolsIndicatorVisible] = (0, _react.useState)(true);
    return /*#__PURE__*/ (0, _jsxruntime.jsx)(DevToolsPopover, {
        routerType: state.routerType,
        semver: state.versionInfo.installed,
        issueCount: errorCount,
        isStaticRoute: state.staticIndicator,
        hide: ()=>{
            setIsDevToolsIndicatorVisible(false);
            fetch('/__nextjs_disable_dev_indicator', {
                method: 'POST'
            });
        },
        setIsErrorOverlayOpen: setIsErrorOverlayOpen,
        isTurbopack: !!process.env.TURBOPACK,
        disabled: state.disableDevIndicator || !isDevToolsIndicatorVisible,
        isBuildError: isBuildError
    });
}
//////////////////////////////////////////////////////////////////////////////////////
const ANIMATE_OUT_DURATION_MS = 200;
const ANIMATE_OUT_TIMING_FUNCTION = 'cubic-bezier(0.175, 0.885, 0.32, 1.1)';
const Context = /*#__PURE__*/ (0, _react.createContext)({});
function getInitialPosition() {
    if (typeof localStorage !== 'undefined' && localStorage.getItem(_shared.STORAGE_KEY_POSITION)) {
        return localStorage.getItem(_shared.STORAGE_KEY_POSITION);
    }
    return INDICATOR_POSITION;
}
function DevToolsPopover(param) {
    let { routerType, disabled, issueCount, isStaticRoute, isTurbopack, isBuildError, hide, setIsErrorOverlayOpen } = param;
    const menuRef = (0, _react.useRef)(null);
    const triggerRef = (0, _react.useRef)(null);
    const turbopackRef = (0, _react.useRef)(null);
    const triggerTurbopackRef = (0, _react.useRef)(null);
    const routeInfoRef = (0, _react.useRef)(null);
    const triggerRouteInfoRef = (0, _react.useRef)(null);
    const preferencesRef = (0, _react.useRef)(null);
    const triggerPreferencesRef = (0, _react.useRef)(null);
    const [isMenuOpen, setIsMenuOpen] = (0, _react.useState)(false);
    const [isTurbopackInfoOpen, setIsTurbopackInfoOpen] = (0, _react.useState)(false);
    const [isRouteInfoOpen, setIsRouteInfoOpen] = (0, _react.useState)(false);
    const [isPreferencesOpen, setIsPreferencesOpen] = (0, _react.useState)(false);
    const [selectedIndex, setSelectedIndex] = (0, _react.useState)(-1);
    const [position, setPosition] = (0, _react.useState)(getInitialPosition());
    // This hook lets us do an exit animation before unmounting the component
    const { mounted: menuMounted, rendered: menuRendered } = (0, _usedelayedrender.useDelayedRender)(isMenuOpen, {
        // Intentionally no fade in, makes the UI feel more immediate
        enterDelay: 0,
        // Graceful fade out to confirm that the UI did not break
        exitDelay: ANIMATE_OUT_DURATION_MS
    });
    const { mounted: turbopackInfoMounted, rendered: turbopackInfoRendered } = (0, _usedelayedrender.useDelayedRender)(isTurbopackInfoOpen, {
        enterDelay: 0,
        exitDelay: ANIMATE_OUT_DURATION_MS
    });
    const { mounted: routeInfoMounted, rendered: routeInfoRendered } = (0, _usedelayedrender.useDelayedRender)(isRouteInfoOpen, {
        enterDelay: 0,
        exitDelay: ANIMATE_OUT_DURATION_MS
    });
    const { mounted: preferencesMounted, rendered: preferencesRendered } = (0, _usedelayedrender.useDelayedRender)(isPreferencesOpen, {
        enterDelay: 0,
        exitDelay: ANIMATE_OUT_DURATION_MS
    });
    // Features to make the menu accessible
    useFocusTrap(menuRef, triggerRef, isMenuOpen);
    useClickOutside(menuRef, triggerRef, isMenuOpen, closeMenu);
    useFocusTrap(turbopackRef, triggerTurbopackRef, isTurbopackInfoOpen);
    useClickOutside(turbopackRef, triggerTurbopackRef, isTurbopackInfoOpen, closeTurbopackInfo);
    useFocusTrap(routeInfoRef, triggerRouteInfoRef, isRouteInfoOpen);
    useClickOutside(routeInfoRef, triggerRouteInfoRef, isRouteInfoOpen, closeRouteInfo);
    useFocusTrap(preferencesRef, triggerPreferencesRef, isPreferencesOpen);
    useClickOutside(preferencesRef, triggerPreferencesRef, isPreferencesOpen, closePreferences);
    function select(index) {
        var _menuRef_current;
        if (index === 'first') {
            var _menuRef_current1;
            const all = (_menuRef_current1 = menuRef.current) == null ? void 0 : _menuRef_current1.querySelectorAll('[role="menuitem"]');
            if (all) {
                const firstIndex = all[0].getAttribute('data-index');
                select(Number(firstIndex));
            }
            return;
        }
        if (index === 'last') {
            var _menuRef_current2;
            const all = (_menuRef_current2 = menuRef.current) == null ? void 0 : _menuRef_current2.querySelectorAll('[role="menuitem"]');
            if (all) {
                const lastIndex = all.length - 1;
                select(lastIndex);
            }
            return;
        }
        const el = (_menuRef_current = menuRef.current) == null ? void 0 : _menuRef_current.querySelector('[data-index="' + index + '"]');
        if (el) {
            setSelectedIndex(index);
            el == null ? void 0 : el.focus();
        }
    }
    function onMenuKeydown(e) {
        e.preventDefault();
        switch(e.key){
            case 'ArrowDown':
                const next = selectedIndex + 1;
                select(next);
                break;
            case 'ArrowUp':
                const prev = selectedIndex - 1;
                select(prev);
                break;
            case 'Home':
                select('first');
                break;
            case 'End':
                select('last');
                break;
            default:
                break;
        }
    }
    function onTriggerKeydown(e) {
        if (isMenuOpen) {
            return;
        }
        // Open with first item focused
        if (e.key === 'ArrowDown' || e.key === 'Enter' || e.key === ' ') {
            e.preventDefault();
            setIsMenuOpen(true);
            // Run on next tick because querying DOM after state change
            setTimeout(()=>{
                select('first');
            });
        }
        // Open with last item focused
        if (e.key === 'ArrowUp') {
            e.preventDefault();
            setIsMenuOpen(true);
            // Run on next tick because querying DOM after state change
            setTimeout(()=>{
                select('last');
            });
        }
    }
    function openErrorOverlay() {
        if (issueCount > 0) {
            setIsErrorOverlayOpen(true);
        }
    }
    function toggleErrorOverlay() {
        setIsErrorOverlayOpen((prev)=>!prev);
    }
    function onTriggerClick() {
        setIsMenuOpen((prev)=>!prev);
    }
    function closeMenu() {
        setIsMenuOpen(false);
        // Avoid flashing selected state
        setTimeout(()=>{
            setSelectedIndex(-1);
        }, ANIMATE_OUT_DURATION_MS);
    }
    function closeTurbopackInfo() {
        setIsTurbopackInfoOpen(false);
    }
    function closeRouteInfo() {
        setIsRouteInfoOpen(false);
    }
    function closePreferences() {
        setIsPreferencesOpen(false);
    }
    function handleHideDevtools() {
        closePreferences();
        hide();
    }
    const [vertical, horizontal] = position.split('-', 2);
    return /*#__PURE__*/ (0, _jsxruntime.jsxs)(_toast.Toast, {
        "data-nextjs-toast": true,
        style: {
            boxShadow: 'none',
            zIndex: 2147483647,
            // Reset the toast component's default positions.
            bottom: 'initial',
            left: 'initial',
            [vertical]: '10px',
            [horizontal]: '20px'
        },
        children: [
            /*#__PURE__*/ (0, _jsxruntime.jsx)(_nextlogo.NextLogo, {
                ref: triggerRef,
                "aria-haspopup": "menu",
                "aria-expanded": isMenuOpen,
                "aria-controls": "nextjs-dev-tools-menu",
                "aria-label": "" + (isMenuOpen ? 'Close' : 'Open') + " Next.js Dev Tools",
                "data-nextjs-dev-tools-button": true,
                disabled: disabled,
                issueCount: issueCount,
                onTriggerClick: onTriggerClick,
                onKeyDown: onTriggerKeydown,
                toggleErrorOverlay: toggleErrorOverlay,
                isDevBuilding: (0, _initialize.useIsDevBuilding)(),
                isDevRendering: (0, _devrenderindicator.useIsDevRendering)(),
                isBuildError: isBuildError
            }),
            routeInfoMounted && /*#__PURE__*/ (0, _jsxruntime.jsx)(_routeinfo.RouteInfo, {
                ref: routeInfoRef,
                routerType: routerType,
                routeType: isStaticRoute ? 'Static' : 'Dynamic',
                isOpen: isRouteInfoOpen,
                setIsOpen: setIsRouteInfoOpen,
                setPreviousOpen: setIsMenuOpen,
                style: {
                    [vertical]: 'calc(100% + 8px)',
                    [horizontal]: 0
                },
                "data-rendered": routeInfoRendered
            }),
            turbopackInfoMounted && /*#__PURE__*/ (0, _jsxruntime.jsx)(_turbopackinfo.TurbopackInfo, {
                ref: turbopackRef,
                isOpen: isTurbopackInfoOpen,
                setIsOpen: setIsTurbopackInfoOpen,
                setPreviousOpen: setIsMenuOpen,
                style: {
                    [vertical]: 'calc(100% + 8px)',
                    [horizontal]: 0
                },
                "data-rendered": turbopackInfoRendered
            }),
            preferencesMounted && /*#__PURE__*/ (0, _jsxruntime.jsx)(_userpreferences.UserPreferences, {
                ref: preferencesRef,
                isOpen: isPreferencesOpen,
                hide: handleHideDevtools,
                setPosition: setPosition,
                position: position,
                style: {
                    [vertical]: 'calc(100% + 8px)',
                    [horizontal]: 0
                },
                "data-rendered": preferencesRendered
            }),
            menuMounted && /*#__PURE__*/ (0, _jsxruntime.jsx)("div", {
                ref: menuRef,
                id: "nextjs-dev-tools-menu",
                role: "menu",
                dir: "ltr",
                "aria-orientation": "vertical",
                "aria-label": "Next.js Dev Tools Items",
                tabIndex: -1,
                className: "dev-tools-indicator-menu",
                onKeyDown: onMenuKeydown,
                "data-rendered": menuRendered,
                style: {
                    '--animate-out-duration-ms': "" + ANIMATE_OUT_DURATION_MS + "ms",
                    '--animate-out-timing-function': ANIMATE_OUT_TIMING_FUNCTION,
                    [vertical]: 'calc(100% + 8px)',
                    [horizontal]: 0
                },
                children: /*#__PURE__*/ (0, _jsxruntime.jsxs)(Context.Provider, {
                    value: {
                        closeMenu,
                        selectedIndex,
                        setSelectedIndex
                    },
                    children: [
                        /*#__PURE__*/ (0, _jsxruntime.jsxs)("div", {
                            className: "dev-tools-indicator-inner",
                            children: [
                                issueCount > 0 && /*#__PURE__*/ (0, _jsxruntime.jsx)(MenuItem, {
                                    title: issueCount + " " + (issueCount === 1 ? 'issue' : 'issues') + " found. Click to view details in the dev overlay.",
                                    index: 0,
                                    label: "Issues",
                                    value: /*#__PURE__*/ (0, _jsxruntime.jsx)(IssueCount, {
                                        children: issueCount
                                    }),
                                    onClick: openErrorOverlay
                                }),
                                /*#__PURE__*/ (0, _jsxruntime.jsx)(MenuItem, {
                                    title: "Current route is " + (isStaticRoute ? 'static' : 'dynamic') + ".",
                                    label: "Route",
                                    index: 1,
                                    value: isStaticRoute ? 'Static' : 'Dynamic',
                                    onClick: ()=>setIsRouteInfoOpen(true),
                                    "data-nextjs-route-type": isStaticRoute ? 'static' : 'dynamic'
                                }),
                                isTurbopack ? /*#__PURE__*/ (0, _jsxruntime.jsx)(MenuItem, {
                                    title: "Turbopack is enabled.",
                                    label: "Turbopack",
                                    value: "Enabled"
                                }) : /*#__PURE__*/ (0, _jsxruntime.jsx)(MenuItem, {
                                    index: 2,
                                    title: "Learn about Turbopack and how to enable it in your application.",
                                    label: "Try Turbopack",
                                    value: /*#__PURE__*/ (0, _jsxruntime.jsx)(ChevronRight, {}),
                                    onClick: ()=>setIsTurbopackInfoOpen(true)
                                })
                            ]
                        }),
                        /*#__PURE__*/ (0, _jsxruntime.jsx)("div", {
                            className: "dev-tools-indicator-footer",
                            children: /*#__PURE__*/ (0, _jsxruntime.jsx)(MenuItem, {
                                "data-preferences": true,
                                label: "Preferences",
                                value: /*#__PURE__*/ (0, _jsxruntime.jsx)(_gearicon.default, {}),
                                onClick: ()=>setIsPreferencesOpen(true),
                                index: isTurbopack ? 2 : 3
                            })
                        })
                    ]
                })
            })
        ]
    });
}
function ChevronRight() {
    return /*#__PURE__*/ (0, _jsxruntime.jsx)("svg", {
        xmlns: "http://www.w3.org/2000/svg",
        width: "16",
        height: "16",
        viewBox: "0 0 16 16",
        fill: "none",
        children: /*#__PURE__*/ (0, _jsxruntime.jsx)("path", {
            fill: "#666",
            fillRule: "evenodd",
            clipRule: "evenodd",
            d: "M5.50011 1.93945L6.03044 2.46978L10.8537 7.293C11.2442 7.68353 11.2442 8.31669 10.8537 8.70722L6.03044 13.5304L5.50011 14.0608L4.43945 13.0001L4.96978 12.4698L9.43945 8.00011L4.96978 3.53044L4.43945 3.00011L5.50011 1.93945Z"
        })
    });
}
function MenuItem(param) {
    let { index, label, value, onClick, href, ...props } = param;
    const isInteractive = typeof onClick === 'function' || typeof href === 'string';
    const { closeMenu, selectedIndex, setSelectedIndex } = (0, _react.useContext)(Context);
    const selected = selectedIndex === index;
    function click() {
        if (isInteractive) {
            onClick == null ? void 0 : onClick();
            closeMenu();
            if (href) {
                window.open(href, '_blank', 'noopener, noreferrer');
            }
        }
    }
    return /*#__PURE__*/ (0, _jsxruntime.jsxs)("div", {
        className: "dev-tools-indicator-item",
        "data-index": index,
        "data-selected": selected,
        onClick: click,
        // Needs `onMouseMove` instead of enter to work together
        // with keyboard and mouse input
        onMouseMove: ()=>{
            if (isInteractive && index !== undefined && selectedIndex !== index) {
                setSelectedIndex(index);
            }
        },
        onMouseLeave: ()=>setSelectedIndex(-1),
        onKeyDown: (e)=>{
            if (e.key === 'Enter' || e.key === ' ') {
                click();
            }
        },
        role: isInteractive ? 'menuitem' : undefined,
        tabIndex: selected ? 0 : -1,
        ...props,
        children: [
            /*#__PURE__*/ (0, _jsxruntime.jsx)("span", {
                className: "dev-tools-indicator-label",
                children: label
            }),
            /*#__PURE__*/ (0, _jsxruntime.jsx)("span", {
                className: "dev-tools-indicator-value",
                children: value
            })
        ]
    });
}
function IssueCount(param) {
    let { children } = param;
    return /*#__PURE__*/ (0, _jsxruntime.jsxs)("span", {
        className: "dev-tools-indicator-issue-count",
        "data-has-issues": children > 0,
        children: [
            /*#__PURE__*/ (0, _jsxruntime.jsx)("span", {
                className: "dev-tools-indicator-issue-count-indicator"
            }),
            children
        ]
    });
}
//////////////////////////////////////////////////////////////////////////////////////
function useFocusTrap(menuRef, triggerRef, isMenuOpen) {
    (0, _react.useEffect)(()=>{
        if (isMenuOpen) {
            var _menuRef_current;
            (_menuRef_current = menuRef.current) == null ? void 0 : _menuRef_current.focus();
        } else {
            var _triggerRef_current, _menuRef_current1;
            const root = (_triggerRef_current = triggerRef.current) == null ? void 0 : _triggerRef_current.getRootNode();
            const activeElement = root instanceof ShadowRoot ? root == null ? void 0 : root.activeElement : null;
            // Only restore focus if the focus was previously on the menu.
            // This avoids us accidentally focusing on mount when the
            // user could want to interact with their own app instead.
            if ((_menuRef_current1 = menuRef.current) == null ? void 0 : _menuRef_current1.contains(activeElement)) {
                var _triggerRef_current1;
                (_triggerRef_current1 = triggerRef.current) == null ? void 0 : _triggerRef_current1.focus();
            }
        }
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [
        isMenuOpen
    ]);
}
//////////////////////////////////////////////////////////////////////////////////////
function useClickOutside(menuRef, triggerRef, isMenuOpen, closeMenu) {
    (0, _react.useEffect)(()=>{
        if (!isMenuOpen) {
            return;
        }
        // Close menu when clicking outside of it or its button
        const handleClickOutside = (event)=>{
            var _menuRef_current, _triggerRef_current;
            if (!(((_menuRef_current = menuRef.current) == null ? void 0 : _menuRef_current.getBoundingClientRect()) ? event.clientX >= menuRef.current.getBoundingClientRect().left && event.clientX <= menuRef.current.getBoundingClientRect().right && event.clientY >= menuRef.current.getBoundingClientRect().top && event.clientY <= menuRef.current.getBoundingClientRect().bottom : false) && !(((_triggerRef_current = triggerRef.current) == null ? void 0 : _triggerRef_current.getBoundingClientRect()) ? event.clientX >= triggerRef.current.getBoundingClientRect().left && event.clientX <= triggerRef.current.getBoundingClientRect().right && event.clientY >= triggerRef.current.getBoundingClientRect().top && event.clientY <= triggerRef.current.getBoundingClientRect().bottom : false)) {
                closeMenu();
            }
        };
        // Close popover when pressing escape
        const handleKeyDown = (event)=>{
            if (event.key === 'Escape') {
                closeMenu();
            }
        };
        document.addEventListener('mousedown', handleClickOutside);
        document.addEventListener('keydown', handleKeyDown);
        return ()=>{
            document.removeEventListener('mousedown', handleClickOutside);
            document.removeEventListener('keydown', handleKeyDown);
        };
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [
        isMenuOpen
    ]);
}
const DEV_TOOLS_INDICATOR_STYLES = "\n  .dev-tools-indicator-menu {\n    -webkit-font-smoothing: antialiased;\n    display: flex;\n    flex-direction: column;\n    align-items: flex-start;\n    background: var(--color-background-100);\n    border: 1px solid var(--color-gray-alpha-400);\n    background-clip: padding-box;\n    box-shadow: var(--shadow-menu);\n    border-radius: var(--rounded-xl);\n    position: absolute;\n    font-family: var(--font-stack-sans);\n    z-index: 1000;\n    overflow: hidden;\n    opacity: 0;\n    outline: 0;\n    min-width: 248px;\n    transition: opacity var(--animate-out-duration-ms)\n      var(--animate-out-timing-function);\n\n    &[data-rendered='true'] {\n      opacity: 1;\n      scale: 1;\n    }\n  }\n\n  .dev-tools-indicator-inner {\n    padding: 6px;\n    width: 100%;\n  }\n\n  .dev-tools-indicator-item {\n    display: flex;\n    align-items: center;\n    padding: 8px 6px;\n    height: var(--size-36);\n    border-radius: 6px;\n    text-decoration: none !important;\n    user-select: none;\n    white-space: nowrap;\n\n    svg {\n      width: var(--size-16);\n      height: var(--size-16);\n    }\n\n    &:focus-visible {\n      outline: 0;\n    }\n  }\n\n  .dev-tools-indicator-footer {\n    background: var(--color-background-200);\n    padding: 6px;\n    border-top: 1px solid var(--color-gray-400);\n    width: 100%;\n  }\n\n  .dev-tools-indicator-item[data-selected='true'] {\n    cursor: pointer;\n    background-color: var(--color-gray-200);\n  }\n\n  .dev-tools-indicator-label {\n    font-size: var(--size-14);\n    line-height: var(--size-20);\n    color: var(--color-gray-1000);\n  }\n\n  .dev-tools-indicator-value {\n    font-size: var(--size-14);\n    line-height: var(--size-20);\n    color: var(--color-gray-900);\n    margin-left: auto;\n  }\n\n  .dev-tools-indicator-issue-count {\n    --color-primary: var(--color-gray-800);\n    --color-secondary: var(--color-gray-100);\n    display: flex;\n    flex-direction: row;\n    align-items: center;\n    justify-content: center;\n    gap: 8px;\n    min-width: var(--size-40);\n    height: var(--size-24);\n    background: var(--color-background-100);\n    border: 1px solid var(--color-gray-alpha-400);\n    background-clip: padding-box;\n    box-shadow: var(--shadow-small);\n    padding: 2px;\n    color: var(--color-gray-1000);\n    border-radius: 128px;\n    font-weight: 500;\n    font-size: var(--size-13);\n    font-variant-numeric: tabular-nums;\n\n    &[data-has-issues='true'] {\n      --color-primary: var(--color-red-800);\n      --color-secondary: var(--color-red-100);\n    }\n\n    .dev-tools-indicator-issue-count-indicator {\n      width: var(--size-8);\n      height: var(--size-8);\n      background: var(--color-primary);\n      box-shadow: 0 0 0 2px var(--color-secondary);\n      border-radius: 50%;\n    }\n  }\n\n  .dev-tools-indicator-shortcut {\n    display: flex;\n    gap: 4px;\n\n    kbd {\n      width: var(--size-20);\n      height: var(--size-20);\n      display: flex;\n      justify-content: center;\n      align-items: center;\n      border-radius: var(--rounded-md);\n      border: 1px solid var(--color-gray-400);\n      font-family: var(--font-stack-sans);\n      background: var(--color-background-100);\n      color: var(--color-gray-1000);\n      text-align: center;\n      font-size: var(--size-12);\n      line-height: var(--size-16);\n    }\n  }\n";

if ((typeof exports.default === 'function' || (typeof exports.default === 'object' && exports.default !== null)) && typeof exports.default.__esModule === 'undefined') {
  Object.defineProperty(exports.default, '__esModule', { value: true });
  Object.assign(exports.default, exports);
  module.exports = exports.default;
}
